### DISCUSSION TITLE
RT apply_doppler and Time Channel simulation

### DISCUSSION THREAD

**Discussion by liuxoB** [NVlabs/sionna#187] [Q&A] [answered]
Dear developers and others, 

I am investigating the effects of the Doppler shift on an arbitrary waveform in time domain. My goal is to simply get the received discrete time signal given the paths and transmit signal. After applying the Doppler shift, calculating the cir and using the `cir_to_time_channel function`, the channel (`hm`) has an extra dimension, 1 for every time step. 

If I then initiate a time channel by using the `ApplyTimeChannel` class and try to use it by calling the instance of the class with my signal and `hm`, I get the error that `ApplyTimeChannel` accepts 1 dimension less. It appears from the documentation that the `num_time_steps` dimension is not accepted. I can only use the channel to get the return signal per time step, which does not seem to be the way I should approach it.

I saw from your tutorial on ray tracing that using the `apply_doppler` and subsequent `.cir` output will directly serve as input to an OFDM channel simulation. This same approach does not works for a time channel simulation.

Is there a way to make this work using sionna?
I look forward to a reply!

Best, Floris

**Comment by jhoydis** [comment, marked as answer]
Hi,

Since you have not provided any code snippet to reproduce your error, I have prepared a quick example showing how you can do time-domain simulations with the channel impulse responses computed by Sionna RT. Please have a look and see how this differs from what you are doing. Hope that helps to clarify your issue.

```
# Imports
%matplotlib inline
import matplotlib.pyplot as plt
import numpy as np
import sionna
from sionna.rt import load_scene, Transmitter, Receiver, PlanarArray
from sionna.channel import ApplyTimeChannel, cir_to_time_channel
from sionna.constants import PI

# Load an example scene and place a transmitter and receiver
scene = load_scene(sionna.rt.scene.simple_street_canyon)

scene.frequency = 3.5e9
scene.tx_array = PlanarArray(num_rows=1,
                             num_cols=1,
                             vertical_spacing=0.5,
                             horizontal_spacing=0.5,
                             pattern="iso",
                             polarization="V")

scene.rx_array = scene.tx_array
scene.add(Transmitter(name="tx",
                      position=[-33,11,32],
                      orientation=[0,0,0]))

scene.add(Receiver(name="rx",
                      position=[27,-14,1.5],
                      orientation=[0,0,0]))


# Compute propagation paths
paths = scene.compute_paths(max_depth=5,
                            num_samples=10e6,
                            diffraction=True)


# Configure parameters need for time-domain link-level simulations
bandwidth=20e6 # bandwidth/sampling frequency
l_min = -6 # Earliest channel tap to consider
l_max = 30 # Last channel tap to consider
l_tot = l_max-l_min+1 # Total length of CIR
num_time_samples = 20000 # Number of input samples
num_time_steps = num_time_samples + l_tot - 1 # Number of channel outputs

# Instantiate ApplyTimeChannel
channel = ApplyTimeChannel(num_time_samples, l_tot, add_awgn=False)

# Create time-varying channel
paths.apply_doppler(bandwidth, num_time_steps, rx_velocities=[10.0, 10.0, 1.0])
a, tau = paths.cir()

# Convert to complex baseband CIR
h = cir_to_time_channel(bandwidth, a, tau, l_min, l_max, normalize=True)

# Visualize CIR
plt.figure()
plt.plot(np.arange(l_min, l_max+1), np.abs(h[0,0,0,0,0,0,:]), "-");
plt.plot(np.arange(l_min, l_max+1), np.abs(h[0,0,0,0,0,10000,:]), "--");
plt.plot(np.arange(l_min, l_max+1), np.abs(h[0,0,0,0,0,20000,:]), "-.");
plt.xlabel("Filter tap index");
plt.ylabel(r"$|h_i|$");
plt.legend([r"$t=0$", r"t=10000", r"t=20000"])

# Create dummy channel input
x = tf.ones([1, 1, 1, num_time_samples], tf.complex64)

# Compute channel output
y = channel([x, h])

# Visualize channel output
plt.figure()
plt.plot(np.abs(y[0,0,0,:]));
plt.xlabel("Time index");
plt.ylabel(r"$|y_i|$");

# Visualize propagation paths
scene.preview(paths=paths)
```

![image](https://github.com/NVlabs/sionna/assets/5190129/7ea855cc-cff1-4728-a285-77a5ab6575a6)
![image](https://github.com/NVlabs/sionna/assets/5190129/a04cd9e2-149f-4644-a162-43d8b709c864)
![image](https://github.com/NVlabs/sionna/assets/5190129/127a7538-77de-463e-800f-57e1d2e4269e)

**Reply by liuxoB** [reply]
Thank you for your fast response! I found the problem, the problem is that I did not have `num_time_steps = num_time_samples + l_tot - 1` , 
I had a different number that I just chose myself. (like `l_max`) Maybe this relation is good to add to the documentation somewhere (or at least I could not find where it was). Other than this my code was roughly the same. Thank you for helping me!

**Reply by jhoydis** [reply]
Great! I think that this is explained in the API documentation of the [`ApplyTimeChannel`](https://nvlabs.github.io/sionna/api/channel.wireless.html#sionna.channel.ApplyTimeChannel) class. But maybe it is not sufficient or not visible enough.

**Reply by liuxoB** [reply]
Ah yes I see it now. I think I didnt make the connection between the sentence "For each batch example, `num_time_samples + l_tot - 1` time steps of a channel realization are required to filter the channel inputs." and the `num_time_steps` parameter.
Thank you for helping me out today, I really appreciate it!
